dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractHostedEntryResource.java

```diff
@@ -0,0 +1,219 @@
+import io.dockstore.webservice.core.Entry;
+import io.dockstore.webservice.core.SourceFile;
+import io.dockstore.webservice.core.Tool;
+import io.dockstore.webservice.core.ToolMode;
+import io.dockstore.webservice.core.User;
+import io.dockstore.webservice.core.Version;
+import io.dockstore.webservice.core.Workflow;
+import io.dockstore.webservice.core.WorkflowMode;
+import io.dockstore.webservice.helpers.ElasticManager;
+import io.dockstore.webservice.helpers.ElasticMode;
+import io.dockstore.webservice.jdbi.EntryDAO;
+import io.dockstore.webservice.jdbi.FileDAO;
+import io.dockstore.webservice.jdbi.UserDAO;
+import io.dockstore.webservice.jdbi.VersionDAO;
+import io.dropwizard.auth.Auth;
+import io.dropwizard.hibernate.UnitOfWork;
+import io.dropwizard.jersey.PATCH;
+import io.swagger.annotations.Api;
+import io.swagger.annotations.ApiParam;
+import org.apache.http.HttpStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Methods to create and edit hosted tool and workflows.
+ * Reuse existing methods to GET them, add labels to them, and other operations.
+ *
+ * @author dyuen
+ */
+@Api("hosted")
+@Produces(MediaType.APPLICATION_JSON)
+public abstract class AbstractHostedEntryResource<T extends Entry<T, U>, U extends Version<U>, W extends EntryDAO<T>, X extends VersionDAO<U>>
+    implements AuthenticatedResourceInterface {
+
+    private static final Logger LOG = LoggerFactory.getLogger(AbstractHostedEntryResource.class);
+    final ElasticManager elasticManager;
+    private final FileDAO fileDAO;
+    private final UserDAO userDAO;
+
+    AbstractHostedEntryResource(FileDAO fileDAO, UserDAO userDAO) {
+        this.fileDAO = fileDAO;
+        this.elasticManager = new ElasticManager();
+        this.userDAO = userDAO;
+    }
+
+    /**
+     * Convenience method to return a DAO responsible for creating T
+     *
+     * @return a DAO that handles T
+     */
+    protected abstract W getEntryDAO();
+
+    /**
+     * Convenience method to return a DAO responsible for creating U
+     *
+     * @return a DAO that handles U
+     */
+    protected abstract X getVersionDAO();
+
+    @POST
+    @Path("/hostedEntry")
+    @Timed
+    @UnitOfWork
+    public T createHosted(@ApiParam(hidden = true) @Auth User user,
+        @ApiParam(value = "For tools, the Docker registry") @QueryParam("registry") String registry,
+        @ApiParam(value = "name", required = true) @QueryParam("name") String name,
+        @ApiParam(value = "Descriptor type", required = true) @QueryParam("descriptorType") String descriptorType) {
+        T entry = getEntry(user, registry, name, descriptorType);
+        long l = getEntryDAO().create(entry);
+        T byId = getEntryDAO().findById(l);
+        elasticManager.handleIndexUpdate(byId, ElasticMode.UPDATE);
+        return byId;
+    }
+
+    protected abstract T getEntry(User user, String registry, String name, String descriptorType);
+
+    @PATCH
+    @io.swagger.jaxrs.PATCH
```

Comment:
Yeah, I think I needed both because swagger didn't seem to pick up the io.dropwizard.jersey.PATCH annotation used to tell Dropwizard that this was a patch. 
It's worth deleting again to see what happens, but the symptom I was seeing was swagger codegen not creating generated methods for this.